src/home_robot/home_robot/mapping/voxel/feature_voxel/conceptfusion.py

```diff
@@ -0,0 +1,662 @@
+            global_feat = self.generate_global_features(img)
+
+            # CLIP features per ROI
+            outfeat = self.generate_local_features(img, masks, global_feat)
+
+            original_intrinsics = scene_obs["intrinsics"][i]
+
+            xyz = unproject_masked_depth_to_xyz_coordinates(
+                depth.unsqueeze(0),
+                camera_pose.unsqueeze(0),
+                original_intrinsics.inverse()[:3, :3].unsqueeze(0),
+            )
+            valid_depth = torch.full_like(
+                depth.squeeze(0), fill_value=True, dtype=torch.bool
+            )
+            if depth is not None:
+                valid_depth = (depth > self.min_depth) & (depth < self.max_depth)
+            valid_depth = valid_depth.flatten()
+            xyz = xyz[valid_depth]
+            features = outfeat.reshape(-1, outfeat.shape[-1])[valid_depth].to(
+                self.device
+            )
+            rgb = (
+                torch.tensor(img, device=self.device)
+                .reshape(-1, 3)[valid_depth]
+                .to(self.device)
+                .float()
+            )
+            self.voxel_map.add(points=xyz, features=features, rgb=rgb)
+
+        return self.voxel_map.get_pointcloud()
+
+    def get_instances_for_queries(self, queries: Sequence[str]):
+        pc_xyz, pc_feat, _, pc_rgb = self.voxel_map.get_pointcloud()
+        instances_dict = {}
+        for class_name in queries:
+            logger.debug("Querying for class: {}".format(class_name))
+            selected_inds, similarity_score = self.query_similarity(class_name, pc_feat)
+
+            if selected_inds.shape[0] <= 1:
+                warnings.warn(
+                    f"No points found for class {class_name}",
+                    UserWarning,
+                )
+                continue
+
+            pc_feat_objects = pc_feat[selected_inds]
+            # pc_rgb_objects = pc_rgb_after_query[selected_inds].squeeze()
+            pc_xyz_objects = pc_xyz[selected_inds].squeeze()
+            similarity_score_objects = similarity_score[selected_inds].squeeze()
+            bounding_boxes, labels = get_bounding_boxes(self.clusterer, pc_xyz_objects)
+
+            instances_per_class = []
+            for idx in range(len(bounding_boxes)):
+                bounds = bounding_boxes[idx]
+
+                volume = float(box3d_volume_from_bounds(bounds).squeeze())
+                min_dim = (bounds[:, 1] - bounds[:, 0]).min()
+
+                if volume < 1e-6 or min_dim < 0.01:
+                    warnings.warn(
+                        f"Skipping box with bounding box {bounds}, volume {volume} and min_dim {min_dim}",
+                        UserWarning,
+                    )
+                else:
+                    instance = Instance(score_aggregation_method="max")
+
+                    # TODO: get rid of the instance view altogether
+                    instance_view = InstanceView(
+                        bbox=None,
```

Comment:
will this get populated if set to None? should it be?

Reply:
It doesn't get populated -- I think the instance view is added just so that we save the embedding. But really we should besetting the pointcloud directly in the Instance.